### PR TITLE
Add baseline ZoneInfo64 data provider

### PR DESCRIPTION
**Pull request by nekevss**
This adds a ZoneInfo64Provider that directly uses zoneinfo64.res bytes. This does not implement any resolution for now for the actual provider.

Ideally, long term, it may be best for this to take in bytes and complete some zerocopy deserialization, but there's some real difficulty around handling zoneinfo64's u16 strings that may require some custom types.

I think this may also require us to carry a Unicode license in the root for `timezone_provider` due to the res files. CC: @Manishearth for confirmation on that

### CODE REVIEW

File: provider/Cargo.toml [hunk 1]
```diff
@@ -34,3 +40,7 @@ databake = { version = "0.2.0", features = ["derive"], optional = true }
 yoke = { version = "0.8.0", features = ["derive"], optional = true }
 serde_json = { version = "1.0.140", optional = true }
 
+# ZoneInfo64 dependencies
+resb = { git = "https://github.com/unicode-org/icu4x.git", optional = true }
+rustc-hash = { workspace = true, optional = true }
+writeable = { workspace = true, optional = true }
```
Comment (HalidOdat) [inline provider/Cargo.toml:46]: Whitespace :policeman: 

```suggestion
writeable = { workspace = true, optional = true }

```

File: provider/src/zoneinfo64.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+#[serde(rename_all = "camelCase")]
+pub struct ZoneTable<'data> {
+    #[serde(borrow)]
+    pub trans_pre32: Option<ZeroVec<'data, i32>>,
+    #[serde(borrow)]
+    pub trans: Option<ZeroVec<'data, i32>>,
+    #[serde(borrow)]
+    pub final_rule: Option<ZoneInfo64String<'data>>,
+    pub final_raw: Option<i32>,
+    pub final_year: Option<i32>,
```
Comment (HalidOdat) [inline provider/src/zoneinfo64.rs:103]: Remark: Seems like from the data that the year can fit in a `i16`

File: provider/src/zoneinfo64.rs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    pub zones: Vec<Zone<'data>>,
+    #[serde(borrow)]
+    pub regions: Vec<ZoneInfo64String<'data>>,
+    // NOTE (nekevss): ZoneInfo64 Rules field is ignored for now. Difficult to deal
+    // with and probably not applicable to resolution anyways.
+}
+
+#[derive(Debug, Clone)]
+pub enum Zone<'data> {
+    Table(ZoneTable<'data>),
```
Comment (HalidOdat) [inline provider/src/zoneinfo64.rs:54]: Since `ZoneTable` is 108 bytes bigger than the `Link` variant maybe we should wrap it in a Box.

```suggestion
    Table(Box<ZoneTable<'data>>),
```

**Review by HalidOdat — APPROVED** [review]
Looks good on my side, just some minor non-blocking nitpicks :smile:

**Comment by Manishearth** [comment]
> I think this may also require us to carry a Unicode license in the root for timezone_provider due to the res files. CC: @Manishearth for confirmation on that

Probably, yes. Make sure it's included in the published package!

File: provider/Cargo.toml [hunk 4]
```diff
@@ -34,3 +40,7 @@ databake = { version = "0.2.0", features = ["derive"], optional = true }
 yoke = { version = "0.8.0", features = ["derive"], optional = true }
 serde_json = { version = "1.0.140", optional = true }
 
+# ZoneInfo64 dependencies
+resb = { git = "https://github.com/unicode-org/icu4x.git", optional = true }
```
Comment (Manishearth) [inline provider/Cargo.toml:44]: We should cut a release; make a PR to ICU4X that finalizes the resb Cargo.toml (make it publish = false), write a changelog entry, and ping me for a release.

File: provider/src/zoneinfo64.rs (hunk 3 above)
Comment (Manishearth) [reply provider/src/zoneinfo64.rs:54]: Not useful if it's being zero-copy deserialized. Could be a Cow though.

**Review by Manishearth — APPROVED** [review]
Yeah, looking at this `ZeroVec` won't help.

We _can_ make it work by adding resb-specific Array, Array16, IntVector types and giving them custom from_bytes behavior.

**Comment by Manishearth** [comment]
More thoughts on zerocopy stuff:


Really I think there's not as much benefit doing this through the serde model if you want to do it zerocopy, doing zerocopy in the serde model is easier if you control the data format itself, but here we do not. If we want to do 100% zerocopy stuff we should write a custom parser with the right intermediate types.

But it's not super necessary, a handful of allocations isn't a big deal here.

File: provider/src/zoneinfo64.rs [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        Self::new()
+    }
+}
+
+impl CompiledZoneInfo64Provider<'static> {
+    pub fn new() -> Self {
+        #[cfg(target_endian = "little")]
+        let bytes = include_bytes!("./data/zoneinfo64/2025b/le/zoneinfo64.res");
+        #[cfg(target_endian = "big")]
+        let bytes = include_bytes!("./data/zoneinfo64/2025b/be/zoneinfo64.res");
```
Comment (Manishearth) [inline provider/src/zoneinfo64.rs:28]: hm. This won't actually work.

ZeroVec does two things: it allows data to be unaligned, and it consistently serializes in little-endian format. You only need the first one if your parsers are different for LE/BE.

Either the resb parser needs to be written to allow selecting endianness in the caller, OR you should exclusively access the ZeroVec data here in a way that inverts the numbers on BE.

Probably something to be fixed later.

Comment (nekevss) [reply provider/src/zoneinfo64.rs:28]: Yeah, I added it initially, and then realized the day after that the BE wouldn't work with ZeroVec. I just haven't gotten around to fixing it just yet.

Comment (robertbastian) [reply provider/src/zoneinfo64.rs:28]: If the `resb` data that is available from platforms uses native endianness, the `resb` crate should just not use ZeroVec.

**Review by robertbastian — COMMENTED** [review]
Would it make sense to move some of the logic, and more importantly the data and tests into the `resb` crate? This should also solve the license problem as `resb` is already Unicode-licensed.

File: provider/src/zoneinfo64.rs (hunk 5 above)
Comment (nekevss) [reply provider/src/zoneinfo64.rs:28]: Agreed. In this case, the zoneinfo file is provided, so the BE zoneinfo file just needs to be removed.

**Comment by nekevss** [comment]
> Would it make sense to move some of the logic, and more importantly the data and tests into the `resb` crate

Yeah, I think there's a good argument here for moving it into the `resb` crate.

The only issue is that I'm not entirely convinced that the `resb` crate should own the zoneinfo64 structs and data. It could probably be feature flagged ... Or maybe an intermediate `zoneinfo64` crate that reexports `resb` and zoneinfo64 data + structs?

**Comment by robertbastian** [comment]
It should definitely be feature-flagged for the scenario where there's already a zoneinfo64 available from somewhere else.

**Comment by nekevss** [comment]
Closing as this is being implemented in a different approach
